src/components/fields/ObjectPropertyField.js

```diff
@@ -0,0 +1,74 @@
+import React, { Component, PropTypes } from "react";
+
+import { getDefaultRegistry } from "../../utils";
+
+class ObjectPropertyField extends Component {
+  static defaultProps = {
+    uiSchema: {},
+    errorSchema: {},
+    idSchema: {},
+    registry: getDefaultRegistry(),
+    disabled: false,
+    readonly: false,
+    autofocus: false,
+    isDefault: true,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = { present: props.required || !props.isDefault };
+  }
+
+  show = event => {
+    event.preventDefault();
+    this.setState({ present: true });
+  };
+
+  hide = event => {
+    event.preventDefault();
+    this.setState({ present: false });
+  };
+
+  render() {
+    const title = this.props.schema.title === undefined
+      ? this.props.name
+      : this.props.schema.title;
+
+    if (this.state.present) {
+      const SchemaField = this.props.registry.fields.SchemaField;
+      const { isDefault, ...props } = this.props;
+      return (
+        <div>
+          <SchemaField {...props} />
+          {this.props.required ||
+            <button onClick={this.hide}>REMOVE {title}</button>}
+        </div>
+      );
+    } else {
+      return <button onClick={this.show}>ADD {title}</button>;
```

Comment:
I'm pretty nervous about the UX in general. I know that @n1k0 took a swing at collapsibility before in https://github.com/mozilla-services/react-jsonschema-form/issues/268 and wasn't able to find a UX that he found felt right for most use cases, and this feels like a similar situation to me. Can we add this only if we are in a circular-schema situation? (N.B. I haven't actually tried running the code so this may not be as intrusive as I imagine it being.)